### ISSUE TITLE
[post-MVP proposal] FMA instruction

### ISSUE THREAD

**Issue by munsocket** [WebAssembly/design#1391]
**Motivation**

Fused multiply–add (FMA) is a floating-point operation performed in one step, with a single rounding. FMA can speed up and improve the accuracy of many computations: dot product, matrix multiplication, convolutions and artificial neural networks, polynomial evaluation, Newton's method, Kahan summation, Veltkamp-Dekker algorithm. This instruction exist in languages like: [C](https://en.cppreference.com/w/c/numeric/math/fma) / [C++](https://en.cppreference.com/w/cpp/numeric/math/fma), [Rust](https://users.rust-lang.org/t/why-does-the-mul-add-method-produce-a-more-accurate-result-with-better-performance/1626), [C#](https://docs.microsoft.com/en-us/dotnet/api/system.math.fusedmultiplyadd?view=net-5.0), [Go](https://golang.org/src/math/fma.go?s=2262:2295), [Java](https://www.geeksforgeeks.org/math-fma-method-in-java-with-examples/), [Julia](https://github.com/JuliaLang/julia/issues/6330), [Swift](https://developer.apple.com/documentation/simd/fma), [OpenGL 4+](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/fma.xhtml).

**Problem and existing solutions**

In WASM there is no way to get speed improvement from this widely supported instruction. There are two way how to implement fallback if hardware is not support it: correctly rounded but slow and simple combination that fast but accumulate error.

For example OpenCL has both. In C/C++ you have to implement fma fallback by yourself but in Rust/Go/C#/Java/Julia fma implemented with correctly rounded fallback. It doesn't make much difference how it will be implemented because you always can detect fma feature in runtime initialisation with special floating point expression and implement conditional fallback as you wish in your software.

If at least first two basic instructions will be implemented it will be great step forward, because right now software that focused on precision need to implement correct fma fallback with more than 23 FLOPs instead of one. This can be finance application or arbitrary precision libraries or space flight simulator with orbital dynamic.

**Proposed instructions**
```lisp
(f64.fma $x $y $z) ;; equivalent to fma(x, y, z) in C99
(f32.fma $x $y $z) ;; equivalent to fmaf(x, y, z) in C99
```
Usually languages compiles fma(x,y,-z) into fused multiply-subtract under the hood. Since .wasm is compilation target looks like all instruction set can be implemented.
```lisp
(f64.fms $x $y $z) ;; RN(x * y - z) = fma(x,y,-z)
(f32.fms $x $y $z) ;;
(f64.fnma $x $y $z) ;; RN(-x * y + z)
(f32.fnma $x $y $z) ;;
(f64.fnms $x $y $z) ;; RN(-x * y - z)
(f32.fnms $x $y $z) ;;
```
But not everyone doing all of them because result with negation the same.

**Implementation**

Here a draft how to implement software fallback based on relatively new [Boldo-Melquiond paper](https://hal-ens-lyon.archives-ouvertes.fr/inria-00080427v2/document).
```c
#include <math.h> //fma, FP_FAST_FMA

static inline double two_prod(const double x, const double y, double &err) {
  double splitter = 0x8000001p0; //2^27+1
  //float splitter = 0x2001p0; //2^13+1

  double t = splitter * a;
  double ah = t + (a - t);
  double al = a - ah;

  t = splitter * b;
  double bh = t + (b - t);
  double bl = b - bh;
  t = a * b;
  
  err = ((ah * bh - t) + ah * bl + al * bh) + al * bl;
  return t;
}

static inline double two_sum(const double a, const double b, double &err){
  double s = a + b;
  double a1  = s - b;

  err = (a - a1) + (b - (s - a1));
  return s;
}

static inline double fma_correct_fallback(const double x, const double y, const double z) {
  // Check overflows
  // Veltkamp-Dekker multiplication: x * y -> (mul, err)
  // Moller-Knuth summation: mul + z -> (sum, err2)
  // Boldo-Melquiond ternary summation: sum + err + err2 -> fma
}

static inline double f64_fma(const double x, const double y, const double z) {
  #ifdef FP_FAST_FMA
    return fma(x, y, z);
  #else
    return fma_correct_fallback(x, y, z);
  #endif
}
```
But [chromium](https://chromium.googlesource.com/external/github.com/WebAssembly/musl/+/landing-branch/src/math/fma.c) and [apple](https://opensource.apple.com/source/Libm/Libm-315/Source/Intel/xmm_fma.c) already had some own implementation.

**Comment by tlively** [comment]
Related: https://github.com/WebAssembly/simd/issues/10 and https://github.com/WebAssembly/simd/pull/79. We'll be looking at how we can best standardize FMA instructions after the current SIMD proposal is finished.

**Comment by arunetm** [comment]
Thanks for proposing this. FMAs are super useful, especially in AI/ML apps. 

This proposal can also pave way for a new class of Scalar & SIMD instructions, ie expanding to sqrt_approximations, reci_sqrt_approximations, float2int conversions in addition to FMA. We have the option to consider introducing a post-MVP "fast-math" option for Wasm Scalar and SIMD including this class of operations that might have similar feature/platform check requirements. 

I have been looking into a few options, we could potentially introduce fast-math option for Wasm in general with a dependency on @tlively's  conditional-section/feature detection proposals. Scalar and SIMD FMA's and the other ops listed above can fit in as the fast-math ops. Does this sound to be a reasonable approach? Happy to help with this effort and extending this to other fast-math operations.

**Comment by moe123** [comment]
Hello, that's a bit more `subtile than what the OP claimed and proposed`. 

```cpp

template <class T>
T fma(const T& x, const T& y, const T& z)
// long double unsupported except if it is an alias to double, or float80; anyways, forget about it, off-topic
{
	const T cs = TwoProduct<T>::split_value; // BTW 2^12 + 1 on float.

	T s = z, w = 0.0, h, q, r, x1, x2, y1, y2;
	
	// can be a loop where x and y are value at index and s is accumulator.
	{
//!# TwoProduct(x[i],y[i],h,r).
	q  = x;
//!# split x[i] into x1,x2.
	r  = cs * q;
	x2 = r - q;
	x1 = r - x2;
	x2 = q - x1;
	r  = y;
//!# h=x[i]*y[i].
	h  = q * r;
//!# split y into y1,y2.
	q  = cs * r;
	y2 = q - r;
	y1 = q - y2;
	y2 = r - y1;
//!# r=x2*y2-(((h-x1*y1) - x2*y1) - x1*y2
	q  = x1 * y1;
	q  = h - q;
	y1 = y1 * x2;
	q  = q - y1;
	x1 = x1 * y2;
	q  = q - x1;
	x2 = x2 * y2;
	r  = x2 - q;
//!# (w,q)=TwoSum(w,h).
	x1 = w + h;
	x2 = x1 - w;
	y1 = x1 - x2;
	y2 = h - x2;
	q  = w - y1;
	q  = q + y2;
	w  = x1;
//!# s=s+(q+r).
	q = q + r;
	s = s + q;
	}
			
	return w + s;
}

```

**Comment by munsocket** [comment]
@moe123 trivial code will fail on overflow test
```c
double expected = fma(DBL_MAX, 2, -DBL_MAX);
double actual = trivial_fallback(DBL_MAX, 2, -DBL_MAX);
```
This is useful (and subtile) if you know that there will be no overflow in code but not suitable for general fallback implementation.

**Comment by mvduin** [comment]
> to my recollection on ARM only M5 series have a full support. That's good for educational purposes, performance and accuracy wise that's a dead-end.

Pretty much every ARM cores that is powerful enough to run a web browser has proper fused multiply-add. Only the first two ARMv7-A cores (Cortex-A8 and -A9) lack it, and all of the ARMv7-R cores with FPU lack it (R4F, R5F, R7F, R8F). Every core since then that has floating-point support includes fused-multiply-add instructions, even including Cortex-M4F microcontrollers.

**Comment by mitschabaude** [comment]
> We'll be looking at how we can best standardize FMA instructions after the current SIMD proposal is finished.

Relaxed SIMD, with `f64x2.relaxed_fma` and similar instructions, was accepted a while ago and has been released in some runtimes.

So, right now we have SIMD versions of FMA but no scalar version i.e. `f64.fma` etc as the OP suggested.

Is it on the roadmap to add scalar versions as well? Or should a new proposal be created for that?
@tlively tagging you as you may know (or may recommend a process)

**Comment by tlively** [comment]
There is not currently a proposal that includes a scalar FMA instruction, so we would need a new proposal for it. We have some general guidance on the process for this [here](https://github.com/WebAssembly/meetings/blob/main/process/proposal.md). Specific interesting questions to answer for a potential FMA instruction would be:

- How widely would it be supported by common hardware?
- Would it be deterministic, or would we have to allow different hardware to produce different results?
- What kind of speedups could we expect on real-world code bases, and how does that vary on different hardware?

**Comment by anonyco** [comment]
It's embarrassing this issue is still open and there's no FMA in JavaScript or in WebAssembly. Moreover, reading through this thread, there seems to be misconceptions about the purpose/use of FMA; let me clear that up too.

FMA is far more essential than a sqrt instruction and is as essential as floating point division to all mathematical operations because of its 106-bit intermediate precision guarantees. All math functions in any half-decent stdlib leverage extensive, heavy usage of FMA in order to provide accurate calculation of transcendental functions to within a ulp accuracy. Without FMA, all stdlib math functions are orders of magnitude slower as they have to carefully balance all manner of intermediate precision extensions like 2sum in order to calculate accurate values for all inputs. (If the stdlib you're using bothers to optimize for non-fma CPUs at all.)

CPUs without FMA are inherently broken and run all stdlib math functions so slowly that its not practical to do more than extremely light floating point arithmetic on them. As such, nobody is using such CPUs for heavy floating point calculations and its insane to optimize/cater WASM to that use-case. 

`relaxed_fma` is only half the solution as its unusable in stdlib math functions because it may be downgraded to multiply+add   and not provide the necessary 106 bits of precision for the math function to fulfill its API contract.

The only correct solution is to provide an FMA SIMD and an FMA scalar and have WASM vendors insert fallback emulation code that gives the correct 106-bit-precise result on CPUs without FMA.  Infact, this is what a lot of math libraries in c/c++ already do. It's hardly worth anyone's time to sink r&d into optimizing for broken CPUs without FMA. 

(This doesn't address the plethora of other systemic issues in WASM SIMD that make it practically unusable for serious applications,  but I digress as that's a battle for another time and just fixing the FMA alone would be a huge step forwards.)

**Comment by tlively** [comment]
@anonyco, you're welcome to help push this along if you'd like. See my previous comment for pointers on next steps.

**Comment by conrad-watt** [comment]
My understanding is that we already deliberately designed the relaxed FMA instruction to facilitate a compilation scheme that "needs" a single-rounding true FMA instruction (especially if we're working under the assumption that it's ok for the code to totally fail to run on machines without native FMA support). I'd be interested to understand if the below already works, without needing a new extension to Wasm.

The procedure:
- Compile the source code to Wasm with the additional assumption that the relaxed FMA Wasm instruction definitely has fast single-rounding behaviour.
- In the client machine, before allowing the compiled Wasm module to run, briefly test whether the Wasm relaxed FMA instruction has single rounding behaviour. If it doesn't, return an error (or load some very slow fallback module).
- If the test passes, it's safe to execute the compiled Wasm module - our "stable" non-determinism guarantees for relaxed SIMD ensure that future FMA executions on the same machine will be single-rounding.

**Comment by sunfishcode** [comment]
@conrad-watt That procedure isn't practical for many users. For example, library authors don't have control over the applications their libraries are used in, so the people writing the code that needs this procedure are not the people in position to implement it. Also, the way one would implement this procedure in browsers isn't portable outside of browsers. Also, implenting and maintaining a fallback path is a burden even in the best cases.

I believe Wasm does want a proper FMA instruction set, and that the main question is just whether we've reached the point where enough CPUs support it.

**Comment by conrad-watt** [comment]
> Also, the way one would implement this procedure in browsers isn't portable outside of browsers. 

I don't think this is totally clear. If necessary, the logic for testing and terminating if support is not detected could be done in pure Wasm within the startup code of the module itself. This would make some forms of fallback more messy, but would be sufficient for users like @anonyco who aren't interested in supporting environments without native FMA.

> That procedure isn't practical for many users. For example, library authors don't have control over the applications their libraries are used in, so the people writing the code that needs this procedure are not the people in position to implement it.
>
> Also, implenting and maintaining a fallback path is a burden even in the best cases.

I'd expect that at least a basic version of this could be done very cheaply at the toolchain level - if you start with a source/compiler-level FMA intrinsic, the bare minimum approach would be to just replace it with the relaxed Wasm instruction and insert a "single-rounding FMA or error" guard on startup. I agree there are scenarios where this wouldn't be sufficient, but I think even something as basic as this would unlock native FMA performance for quite a lot of users relatively quickly, compared to a standards proposal.

I don't think the library scenario is uniquely problematic. I'd expect that library source code would normally just use the relevant source-level FMA primitive, so the most common Wasm compilation scenario (where the library and user code are compiled together under the same configuration) would "just work". Even for separately-compiled-to-Wasm libraries, I don't think the possible behaviours (in the worst, mis-matched cases) are much weirder than we'd get anyway if we added an always-single-rounding FMA instruction to Wasm and allowed compilation to it - the main thing you're getting from the instruction is an engine-implemented fallback, but even without this the library is free to decide if it wants to supply its own software (toolchain-generated?) fallback, or just error immediately.

**Comment by hanna-kruppe** [comment]
Using relaxed-SIMD FMA with a runtime check is a neat trick for mostly getting the desired codegen without wasm engine updates, but it doesn't lead to a desirable state of affairs long-term. Hopefully, eventually, every relevant engine should support *scalar* single-rounding FMA, so there won't be any need for fallbacks anymore. In contrast, not every non-browser engine is eager to implement the SIMD extension, let alone relaxed-SIMD. Even non-relaxed SIMD is a huge surface area to implement, significantly increase the footprint of engines that optimize for that metric[^1], and probably won't help performance if you're either interpreting wasm[^2] or compiling to native code for a CPU without 128-bit SIMD (e.g., any RISC-V core without the vector extension).

For comparison, the ["Lime" subset of wasm](https://github.com/WebAssembly/tool-conventions/blob/main/Lime.md) exists for engines that will likely never implement reference types or GC, and specifically includes some useful bits of the reference-types and bulk-memory proposals that is palatable for those engines (in particular, doesn't require them to actually implement any reference *types*). For FMA, there's no obvious subset of relaxed-SIMD that an engine could implement. Even if you try to subset instructions, you need to support the v128 type to begin with and some way to convert to and from scalars, and then someone needs to actually define what that subset is so toolchains and engines can agree on it.

[^1]: Even wasmtime, which has implemented relaxed SIMD (and GC! and exception handling! and most other features!), also supports configurations without SIMD to better accommodate embedded use cases: https://github.com/bytecodealliance/wasmtime/pull/10727
[^2]: in theory having fewer opcodes that do more work is nice for an interpreter, but having to support v128 values everywhere has costs as well, and some operations like shuffles are fast in hardware but relatively slow in software unless your interpreter can itself use native SIMD.

**Comment by conrad-watt** [comment]
Sorry - you're right that I'm not thinking properly about the value of a scalar FMA instruction independent from SIMD.

If we tried to go down the path of supporting scalar FMA, would the proportion of supporting hardware look similar to SIMD FMA, or is the story more complicated?

**Comment by hanna-kruppe** [comment]
Hardware with scalar FMA instructions but no SIMD exist, though it's mostly confined to a certain slice of embedded SoCs. Anything that's called an "application processor" or doing serious signal processing is very likely to have SIMD, and anything *really* tiny or simple will probably not have an FPU at all. There's a middle ground in between but I'm not sure how large the intersection with "someone wants to run wasm on it" is.

For example, Arm Cortex-M55 and successors have options for 128-bit SIMD (Helium / MVE) and a scalar FPU. Both the FPU and floating-point SIMD are optional (i.e., you can buy cores with and without them) but the SIMD only supports f32 elements at most (no f64 option) while the scalar FPU goes up to f64 when it's included.

In RISC-V land, cores with full hardware floating-point but no SIMD were pretty common (e.g., the initial target for many Linux distros was `rv64gc` which has scalar FMA but no SIMD of any kind) before the vector extension was ratified and put in [RVA23](https://docs.riscv.org/reference/profiles/rva23/_attachments/rva23-profile.pdf) hardware profile. I can't tell if that will remain true going forward, but at least they also ratified another hardware profile at the same time ([RVB23](https://docs.riscv.org/reference/profiles/rvb23/_attachments/rvb23-profile.pdf)) that requires scalar floating-point (including FMA) but keeps vectors optional. There's also enough demand for "can afford an FPU but can't afford an extra 32 registers" that RISC-V [standardized a way](https://docs.riscv.org/reference/isa/unpriv/zfinx.html) to implement an FPU while reusing the integer registers.

**Comment by l1mey112** [comment]
> My understanding is that we already deliberately designed the relaxed FMA instruction to facilitate a compilation scheme that "needs" a single-rounding true FMA instruction (especially if we're working under the assumption that it's ok for the code to totally fail to run on machines without native FMA support). I'd be interested to understand if the below already works, without needing a new extension to Wasm.

@conrad-watt this procedure already works for my library RandomX.js ([detect.ts](https://github.com/l1mey112/randomx.js/blob/master/src/detect/detect.ts), [fma.wat](https://github.com/l1mey112/randomx.js/blob/master/src/detect/fma.wat)) but it's hardly optimal. I'm relying on the fact that the tens of thousands of compilations after the detect will have the same single rounding behaviour as in the detect stage. The FMA is for performance but since it's used in a hash function it needs to be binary-equal to an actual single round.

In my opinion, the relaxed SIMD specification giving us faulty FMAs then stopping there is a terrible idea.

I got lucky here because I JIT my WASM binaries I can choose whether or not I want to copy in the FMA version or the normal slow version, and I just do so ([jit_vm.c](https://github.com/l1mey112/randomx.js/blob/3156507edab9fa1545a4406a489e7d03474d9241/src/jit/jit_vm.c#L755)), but for libraries they're out of luck to do this in a nice way.

**Comment by l1mey112** [comment]
As I understand it, the core 5 **IEEE-754 2008** instructions that must be properly rounded is addition, subtraction, multiplication, division, square root, **AND** fused multiply-add. Since we have `f64.(add|sub|mul|div|sqrt)`, why not `f64.fma` ?

**Comment by anonyco** [comment]
I took a break for a week ha ha. @tlively I looked at those two threads and both seem to deviate from the kind of FMA that’s needed in WASM. I’d be happy to help in any way to push this along but I’m not going to invest too much effort as (1.) a proper FMA won’t fix all the other issues with WASM and (2.) it seems likely from these threads that WASM is heading in the direction of a poor fma implantation that’s going to be useless in practice.    

OK, my response to everyone (@tively @hanna-kruppe @conrad-watt) is the same: have you ever written a compiler, audited an ABI specification, or ported a toolchain to a new platform? If you have experience with at least one of those (I have all three), you’d understand how strict only-if-supported FMA that throws errors if the hardware lacks support is a non-option as we’d see the exact same thing happen as we see today in x86_64: the overwhelming majority of numeric C/C++ code runs extremely slowly on x86_64 due to libc calls necessitated by the extremely lacking SSE/2 capabilities of the baseline instruction set. It’d be the same deal with throwing upon unsupported WASM: it’d benefit almost no applications as compilers would be forced to compile in a libc call. Then, high performance WASM projects with the know-how  to do runtime WASM dispatch would be seen implementing fma with a libc function call anyway as non-fma CPUs are few and far between.

**TL;DR:** There is one and only one way to correctly implement FMA in WASM: have vendors insert slow fma libc call fallbacks to emulate fma on processors without FMA support. Relaxed FMA and only-if-supported FMA are nice ideas from a hypothetical standpoint, but I can guarantee neither will work well in practice.

Notice: the relaxed FMA proposal has two possible outcomes if it were approved: (A.) the first option is it causes too many toolchain issues for it to ever see widespread usage. (B.) it does receive widespread use but then causes issues with WASM runtimes like web browsers giving wrong results or possibly erroring (due to bad application design) on hardware without FMA, forcing some vendors to emulate relaxed fma with libc calls, which fractures the ecosystem as other vendors won’t, causing relaxed fma to become inconsistent and moot to actually use. There is no option C where relaxed FMA magically works out happily ever after; that’s not how toolchains and ABIs work.

@l1mey112 I think you misunderstand the semantics of what rounding means. Let `RN` round an infinite precision float to 53 bits of precision. Then, multiply+add of x and y and z is `RN(RN(x,y)+z)`. In comparison, FMA of x and y and a is `RN(x*y+z)`

**Comment by tlively** [comment]
No need to demean the work of everyone here. We're all working under many competing design constraints, and of course there are many other projects we're all prioritizing as well.

Wasm has historically avoided adding instructions that need slow software implementations on common consumer CPUs. This is because "predictable performance" has long been a design goal. That's why your preferred solution is only a good fit for Wasm if the vast majority of consumer CPUs currently in use support scalar FMA.

Some cursory research shows that Intel/AMD have supported scalar FMA for at least a decade and ARM supports it since ARMv8, so we might be good on the hardware support front.

The other motivating questions to answer are about potential users. For example, what kind of user would benefit from scalar FMA but cannot use the existing v128.relaxed_multiply_add? How much would they benefit?

**Comment by anonyco** [comment]
Rereading my reply, I recognize I sounded condescending and that was not my intention. My intention was to be pragmatic about software design. I see the theoretical merits of everyone else’s proposals but I know for an indisputable fact they won’t work in practice. I can provide many related historical incidences to back every assertion I made if you need proof.

I can answer your later question: all users of all software will immediately benefit from scalar fma with guaranteed fallback as the relaxed fma is useless for precision guarantees in the libc and significantly slows down all libc math calls. No, calling JavaScript math from WASM isn’t a solution at all.

**Comment by hanna-kruppe** [comment]
> The other motivating questions to answer are about potential users. For example, what kind of user would benefit from scalar FMA but cannot use the existing v128.relaxed_multiply_add? How much would they benefit?

To a first approximation, every user who calls into non-trivial libm functions stands to benefit indirectly, though the benefit would probably be quite small for most of them. Correctly rounded FMA is a very useful building block both for efficiency and accuracy of many libm functions. For example, CORE-MATH seems to use it for many trig functions some versions of log/exp, erf (binary64 only) and cbrt (binary64 only). Performance on targets without FMA is usually not a priority these days, and the correctness/accuracy proofs often heavily rely on correctly rounded FMA, so it's common to use a software fallback instead of a totally different implementation that avoids FMA.

I don't have time or resources to collect hard numbers here, but qualitatively speaking, for the libm use case:

1. Even users that *can* assume relaxed SIMD is available will usually need runtime detection for single vs double rounding. This has either runtime cost (if done within the wasm module) or requires selecting between multiple variants of your module(s) or components at instantiation time, which causes various problems as mentioned above (and most users probably won't want to deal with that pain for small gains).
2. As mentioned above, some embedded systems either don't have the necessary SIMD instructions or don't want to pay the engine footprint cost of implementing Webassembly SIMD. I expect that some of the users use libm functions occasionally, even if they're not doing heavy duty number crunching that would really benefit from SIMD.

For the first case, one thing that could be done today, without a wasm proposal, is to implement this runtime detection in the implementation of `fma`/`fmaf` or closest equivalent (not at instantiation time). Native code sometimes does this for FMA instruction availability and it's a great performance boost there vs. always using the software fallback (see e.g. https://github.com/rust-lang/compiler-builtins/pull/896 for a recent example with microbenchmarks).

However, such runtime detection is more expensive than statically knowing that FMA is available. I'd also expect that the relative cost is higher for wasm than for native code, since there is no equivalent to e.g. GNU ifuncs and the relative cost of direct vs indirect call is generally a bit higher for wasm than for native (ignoring speculative inlining, which most engines don't do). So if scalar single-rounded FMA was widely supported, this would be preferable for math libraries.

An interesting experiment would be to implement the runtime detection approach in e.g. wasi-libc or Rust's compiler-builtins and do some microbenchmarks to evaluate:

- If you do the within-wasm runtime check for single vs double rounding, how much faster is this than the scalar software fallbacks used today? Is there still a gap vs. native libm code doing runtime detection for availability of a native FMA instruction?
- If you don't do any runtime check (i.e., just assume that it's single-rounding), how much does that improve performance? That's the performance gain users will see if they can't don't want to ship two variants and do instantiation-time selection.
- With current engines, does the detour through SIMD to get scalar FMA have a cost that a scalar instruction would avoid? This may be hard to measure without engine modifications, but if there is some gap with native performance observed in the previous benchmarks, staring the at assembly generated by current engines would at least reveal if adding peephole optimizations for this pattern *could* help.

There is also a (very minor) code size cost to putting a copy of the software fallback in every module/components that needs it. It's probably negligible compared to average module size, but it might be mildly unsatisfying if you have a lot of different modules or many small components.

**Comment by anonyco** [comment]
> so it's common to use a software fallback instead of a totally different implementation that avoids FMA.

Bingo! Right on the money: there is no reason not to provide a software fallback for FMA because (1.) anybody using a 20 year old embedded CPU without FMA is painfully aware of how limited their CPU is and is avoiding floating points, and (2.) most (all?) libc (as well as most software that uses fma in general) handles CPUs without fma via software emulation of it anyway.

> However, such runtime detection is more expensive than statically knowing that FMA is available.

There’s a bigger picture at play: most users don’t have the time/resources/awareness to waste a few hours to a day to delve into runtime detection and specialization of their software for a feature that should be the default—the FMA fallback. As a consequence, almost no users do and everyone looses greatly.

> Even users that can assume relaxed SIMD is available will usually need runtime detection for single vs double rounding. 

My understand was the main users of relaxed SIMD are those who compile with `-ffast-math` (enabled by default at `-Ofast` if that’s more familiar.) Correct me if I’m wrong.

> If you do the within-wasm runtime check for single vs double rounding, how much faster is this than the scalar software fallbacks used today?

When I benchmarked this a few years ago, it was 25-35x slower latency and 100-150x worse throughout depending on the CPU/library I tried. I’d expect these numbers to be even worse today with more fma ports on the typical cpu. E.g. Musl libc implements their FMA fallback using integer arithmetic, which is actually probably one of the saner ways to handle the full battery of required s/q NAN behavior and errno *correctly*. (Albeit very very slowly.)

> With current engines, does the detour through SIMD to get scalar FMA have a cost that a scalar instruction would avoid? 

Exactly 33-40% worse latency and 33-50% worse throughput for the majority of CPUs to unpack double-precision SIMD and repack it to use scalar FMA as the swizzling operations compete with the FMA’s execution ports. (For single precision, it’d be 56-63% worse latency and 63-67% worse throughout assuming the minimum 3 instructions to unpack and 3 to repack.)

Granted, it wouldn’t be the worse thing as WASM SIMD already has so many fatal design flaws impeding efficient utilization of hardware that those numbers would be amortized by all the other slowdowns in the SIMD code.

> There is also a (very minor) code size cost to putting a copy of the software fallback in every module/components that needs it.

From what I’ve seen, even the most complex FMA fallbacks take up extremely insignificant space as they don’t need lookup tables, switches, or a lot of global constant relocations.

**Comment by mvduin** [comment]
> ARM supports it since ARMv8

All but the oldest ARMv7 cores support FMA as well.

**Comment by anonyco** [comment]
Note: ARMv7 is its own special soup mix mess as its VFP supports both single and double precision but ARMv7 NEON only supports single precision. ARMv7 single precision neon also lacks a IEEE compliance mode: https://developer.arm.com/documentation/dui0204/h/Bcfhfbga

That same document shows the fmac fused multiply accumulate double precision FPU instruction documented as far back as 2002 in the VFPv2 of ARMv5TE, ARMv5TEJ and ARMv6: https://developer.arm.com/documentation/dui0204/g/neon-and-vfp-programming/vfp-instructions/fmac--fnmac--fmsc--and-fnmsc?lang=en

The real issue with ARMv7 and other 32-bit arm is that the VFP isn’t a requirement. If it’s absent, software emulated floating points must be employed. Notice that software emulated FMA should be about the same speed as software emulated fmul/fadd, so we don’t have to worry about ARMv7 CPUs.

**Comment by jakobkummerow** [comment]
@hanna-kruppe :
> With current engines, does the detour through SIMD to get scalar FMA have a cost that a scalar instruction would avoid? This may be hard to measure without engine modifications

To make the benefits of scalar FMA instructions measurable, I'd be open to implementing them in V8, or reviewing a patch if someone wants to beat me to it. (Of course, that implementation would be behind an experimental flag as long as it's not a finalized proposal.)
That said, an engine-only prototype is likely of limited use, unless someone else volunteers to implement support in some Wasm-producing toolchain (probably Emscripten/LLVM would be best, to support compiling libc?).
To coordinate prototyping efforts, a draft spec would be useful. In particular, it should specify a binary encoding (`0xFC 0xYZ`?), and a few correctness tests, and for bonus points perhaps even a reference fallback implementation. As far as I'm concerned, an informal document is enough to get started.

**Comment by anonyco** [comment]
**EDIT:** It seems I submitted an old revision of my rfc and can't find the latest; I'll get around to fixing this soon.
